### PR TITLE
8265079: Implement VarHandle invoker caching

### PR DESCRIPTION
**Pull request by JornVernee**
This patch implements 2 leftover TODOs for implementing var handle invoker MH caching (lambda forms for those were already shared/cached).

This piggybacks on the existing mechanism for method handle invoker caching.

Testing: Local testing `java/lang/invoke` tests. Tier 1-3

Thanks,
Jorn

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265079](https://bugs.openjdk.java.net/browse/JDK-8265079): Implement VarHandle invoker caching

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to fa5a721f6daacd81ce50dd878ca162fa3da0f373
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to df10dbddb57ae706cbcc3c43873ae736f2c86dff
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to fa5a721f6daacd81ce50dd878ca162fa3da0f373
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to fa5a721f6daacd81ce50dd878ca162fa3da0f373

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3439/head:pull/3439` \
`$ git checkout pull/3439`

Update a local copy of the PR: \
`$ git checkout pull/3439` \
`$ git pull https://git.openjdk.java.net/jdk pull/3439/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3439`

View PR using the GUI difftool: \
`$ git pr show -t 3439`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3439.diff">https://git.openjdk.java.net/jdk/pull/3439.diff</a>

</details>

### CODE REVIEW

**Review by cl4es — APPROVED** [review]
LGTM

`MethodHandles.varHandle(Exact)Invoker` might be somewhat obscure, but caching them like this is straightforward and basically free.

File: src/java.base/share/classes/java/lang/invoke/Invokers.java [hunk 1]
```diff
@@ -122,6 +128,14 @@ private synchronized MethodHandle setCachedInvoker(int idx, final MethodHandle i
         return invokers[idx] = invoker;
     }
 
+    private MethodHandle cachedVHInvoker(int idx, VarHandle.AccessMode ak) {
```
Comment (iwanowww) [inline src/java.base/share/classes/java/lang/invoke/Invokers.java:131]: You can turn `int idx` parameter into `boolean isExact` and choose base index depending on its value.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by JornVernee** [comment]
Thanks for the reviews. Comment addressed.

File: test/jdk/java/lang/invoke/TestVHInvokerCaching.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            long f_long;
+            float f_float;
+            double f_double;
+            Object f_Object;
+            String f_String;
+        }
+
+        MethodHandles.Lookup lookup = lookup();
+
+        for (Class<?> type : TEST_TYPES) {
```
Comment (PaulSandoz) [inline test/jdk/java/lang/invoke/TestVHInvokerCaching.java:88]: A simpler approach would be to iterate over the fields of `Holder` and use `unreflectVarHandle`, then you can remove `TEST_TYPES`.

Comment (JornVernee) [reply test/jdk/java/lang/invoke/TestVHInvokerCaching.java]: Yeah, that's a good idea. Fixed

**Review by cl4es — APPROVED** [review]

**Review by PaulSandoz — APPROVED** [review]

File: test/jdk/java/lang/invoke/TestVHInvokerCaching.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8265079
+ * @run testng/othervm -Xverify:all -ea -esa TestVHInvokerCaching
```
Comment (mlchung) [inline test/jdk/java/lang/invoke/TestVHInvokerCaching.java:26]: Nit: the makefile to run jtreg set `-ea -esa`.   It's not strictly necessary in `@run`.  If you want to run it standalone, you can run with `jtreg -ea -esa` option.

**Review by mlchung — APPROVED** [review]

File: test/jdk/java/lang/invoke/TestVHInvokerCaching.java (hunk 3 above)
Comment (JornVernee) [reply test/jdk/java/lang/invoke/TestVHInvokerCaching.java:26]: Ok, I will remove them.

**Pull request closed without merging**